mvpa2/tests/test_group_clusterthr.py

```diff
@@ -0,0 +1,303 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the PyMVPA package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Unit tests for Stelzer et al. cluster thresholding algorithm"""
+
+from mvpa2.base import externals
+from mvpa2.testing.tools import SkipTest
+
+# TODO a tiny bit also needs statsmodels
+if not externals.exists('scipy'):
+    raise SkipTest
+
+from collections import Counter
+import numpy as np
+import random
+
+from mvpa2.testing import assert_array_equal, assert_raises, assert_equal, \
+    assert_array_almost_equal, assert_almost_equal, assert_true, assert_false
+import mvpa2.algorithms.group_clusterthr as gct
+from mvpa2.datasets import Dataset, dataset_wizard
+from mvpa2.mappers.base import IdentityMapper
+from nose.tools import assert_greater_equal, assert_greater
+from mvpa2.testing.sweep import sweepargs
+
+from scipy.ndimage import measurements
+from scipy.stats import norm
+
+def test_pval():
+    def not_inplace_shuffle(x):
+        x = list(x)
+        random.shuffle(x)
+        return x
+
+    x = range(100000) * 20
+    x = np.array(x)
+    x = x.reshape(20, 100000)
+    x = x.T
+    x = np.apply_along_axis(not_inplace_shuffle, axis=0, arr=x)
+    expected_result = [100000-100000*0.001]*20
+
+    thresholds = gct.get_thresholding_map(x, p=0.001)
+    assert_array_equal(thresholds, expected_result)
+    # works with datasets too
+    dsthresholds = gct.get_thresholding_map(Dataset(x), p=0.001)
+    assert_almost_equal(thresholds, dsthresholds)
+    assert_raises(ValueError,
+                  gct.get_thresholding_map, x, p=0.00000001)
+
+    x = range(0,100,5)
+    null_dist = np.repeat(1, 100).astype(float)[None]
+    pvals = gct._transform_to_pvals(x, null_dist)
+    desired_output = np.array([1, 0.95, 0.9, 0.85, 0.8, 0.75, 0.7, 0.65, 0.6,
+                      0.55, 0.5, 0.45, 0.4, 0.35, 0.3, 0.25, 0.2, 0.15, 0.1,
+                      0.05])
+    assert_array_almost_equal(desired_output, pvals)
+
+
+def test_cluster_count():
+    # we get a ZERO cluster count of one if there are no clusters at all
+    # this is needed to keept track of the number of bootstrap samples that yield
+    # no cluster at all (high treshold) in order to compute p-values when there is no
+    # actual cluster size histogram
+    assert_equal(gct._get_map_cluster_sizes([0,0,0,0]), [0])
+    # if there is at least one cluster: no ZERO count
+    assert_equal(gct._get_map_cluster_sizes([0,0,1,0]), [1])
+    for i in range(2):  # rerun tests for bool type of test_M
```

Comment:
I would recommend to make it a "parametric" test where  `i` is an argument.  Two possible ways

1.  use our sweepargs```@sweepargs(i=[1,2])```

2. making it "nose parametric":
 a. rename this one to e.g. _test_cluster_count(i)
 b. make
```python
def test_cluster_count():
       for i in [1,2]:
           yield _test_cluster_count, i
```

imho we should start using 2 more often ;)